pkg/warning/warning.go

```diff
@@ -0,0 +1,111 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package warning contains utilities for handling warnings.
+package warning
+
+import (
+	"fmt"
+	"regexp"
+	"sync"
+
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+)
+
+// LoggingHandler is a wrapper around a provided logr.Logger that implements the
+// rest.WarningHandler interface.
+type LoggingHandler struct {
+	// logger is used to log responses with the warning header
+	logger logging.Logger
+	// deduplicate indicates a given warning message should only be written once.
+	deduplicate bool
+	// discardWarnings controls whether warnings should be discarded.
+	discardWarnings []regexp.Regexp
+
+	// writtenLock gurads written
+	writtenLock sync.Mutex
+	// used to keep track of already logged messages
+	// and help in de-duplication.
+	written map[string]struct{}
+}
+
+// HandleWarningHeader handles logging for responses from API server that are
+// warnings with code being 299 and uses a logr.Logger for its logging purposes.
+func (l *LoggingHandler) HandleWarningHeader(code int, _ string, message string) {
+	if code != 299 || len(message) == 0 {
+		return
+	}
+
+	for _, discard := range l.discardWarnings {
+		if discard.MatchString(message) {
+			return
+		}
+	}
+
+	if l.deduplicate {
+		l.writtenLock.Lock()
+		defer l.writtenLock.Unlock()
+
+		if _, alreadyLogged := l.written[message]; alreadyLogged {
+			return
+		}
+		l.written[message] = struct{}{}
+	}
+	l.logger.Info(message)
+}
+
+// LoggingHandlerOption is used to configure a LoggingHandler.
+type LoggingHandlerOption func(*LoggingHandler)
+
+// Deduplicate is an option that can be used to enable de-duplication of
+// warning messages.
+func Deduplicate() LoggingHandlerOption {
+	return func(l *LoggingHandler) {
+		l.deduplicate = true
+	}
+}
+
+// DiscardRegexp is an option that can be used to specify a regular
+// expression that matches warnings that should be discarded.
+func DiscardRegexp(r regexp.Regexp) LoggingHandlerOption {
+	return func(l *LoggingHandler) {
+		l.discardWarnings = append(l.discardWarnings, r)
+	}
+}
+
+// NewLoggingHandler returns an implementation of rest.WarningHandler that
+// logs warnings with code = 299 to the provided logr.Logger.
+func NewLoggingHandler(l logging.Logger, opts ...LoggingHandlerOption) *LoggingHandler {
+	k := &LoggingHandler{logger: l}
+
+	for _, opt := range opts {
+		opt(k)
+	}
+
+	if k.deduplicate {
+		k.written = map[string]struct{}{}
+	}
+	return k
+}
+
+// DomainQualifiedFinalizer returns a regular expression that matches warnings
+// that suggest a domain-qualified finalizer name should be used to avoid
+// accidental conflicts with other finalizer writers.
+func DomainQualifiedFinalizer(domain string) regexp.Regexp {
+	return *regexp.MustCompile(
+		fmt.Sprintf("^metadata.finalizers:.*%s.*prefer a domain-qualified finalizer name to avoid accidental conflicts with other finalizer writers$", domain),
```

Comment:
will this check start to fail once we pick up https://github.com/kubernetes/kubernetes/pull/127303/files which changes the warning message?